test_conformance/api/negative_device.cpp

```diff
@@ -0,0 +1,509 @@
+                              nullptr, &number_of_properties);
+    test_error(err, "clGetDeviceInfo");
+    std::vector<cl_device_partition_property> supported_properties(
+        number_of_properties / sizeof(cl_device_partition_property));
+    err = clGetDeviceInfo(deviceID, CL_DEVICE_PARTITION_PROPERTIES,
+                          number_of_properties, &supported_properties.front(),
+                          nullptr);
+    test_error(err, "clGetDeviceInfo");
+    int ret = SupportedPartitionSchemes::None;
+    for (auto property : supported_properties)
+    {
+        switch (property)
+        {
+            case CL_DEVICE_PARTITION_EQUALLY:
+                ret |= SupportedPartitionSchemes::Equally;
+                break;
+            case CL_DEVICE_PARTITION_BY_COUNTS:
+                ret |= SupportedPartitionSchemes::Counts;
+                break;
+            case CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN:
+                ret |= SupportedPartitionSchemes::Affinity;
+                break;
+            default: break;
+        }
+    }
+    return ret;
+}
+
+static std::vector<cl_device_partition_property>
+get_invalid_properties(int unsupported_properties)
+{
+    if (unsupported_properties & SupportedPartitionSchemes::Equally)
+    {
+        return { CL_DEVICE_PARTITION_EQUALLY, 1, 0 };
+    }
+    else if (unsupported_properties & SupportedPartitionSchemes::Counts)
+    {
+        return { CL_DEVICE_PARTITION_BY_COUNTS, 1, 0 };
+    }
+    else if (unsupported_properties & SupportedPartitionSchemes::Affinity)
+    {
+        return { CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN,
+                 CL_DEVICE_AFFINITY_DOMAIN_NEXT_PARTITIONABLE, 0 };
+    }
+    else
+    {
+        return {};
+    }
+}
+
+static cl_uint get_uint_device_info(const cl_device_id deviceID,
+                                    const cl_device_info param_name)
+{
+    cl_uint ret = 0;
+    cl_int err =
+        clGetDeviceInfo(deviceID, param_name, sizeof(ret), &ret, nullptr);
+    test_error(err, "clGetDeviceInfo");
+    return ret;
+}
+
+int test_negative_create_sub_devices(cl_device_id deviceID, cl_context context,
+                                     cl_command_queue queue, int num_elements)
+{
+    int supported_properties = get_supported_properties(deviceID);
+    if (supported_properties == SupportedPartitionSchemes::None)
+    {
+        printf("Device does not support creating subdevices... Skipping\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    cl_device_partition_property properties[3] = {};
+    cl_uint max_compute_units =
+        get_uint_device_info(deviceID, CL_DEVICE_MAX_COMPUTE_UNITS);
+    cl_uint max_sub_devices =
+        get_uint_device_info(deviceID, CL_DEVICE_PARTITION_MAX_SUB_DEVICES);
+    if (supported_properties & SupportedPartitionSchemes::Equally)
+    {
+        properties[0] = CL_DEVICE_PARTITION_EQUALLY;
+        properties[1] = max_compute_units;
```

Comment:
I don't think this is a valid partitioning since there will only be one output device.

Reply:
This is still an issue that needs to be solved.

Our CPU device is currently "failing" this test because this partitioning is invalid - it only creates one output device.  Since we check for this condition relatively early we return `CL_DEVICE_PARTITION_FAILED` in the case where `num_devices` is zero rather than `CL_INVALID_VALUE:

```
>>>> clCreateSubDevices: in_device = Intel(R) Core(TM) i5-7300U CPU @ 2.60GHz (CL_DEVICE_TYPE_CPU) (0x558e839082b8), properties = [ CL_DEVICE_PARTITION_EQUALLY = 4 ], num_devices = 0
<<<< clCreateSubDevices -> CL_DEVICE_PARTITION_FAILED
ERROR: clCreateSubDevices should return CL_INVALID_VALUE when: "out_devices is not NULL and num_devices is less than the number of sub-devices created by the partition scheme"! (Got CL_DEVICE_PARTITION_FAILED, expected CL_INVALID_VALUE from /home/bashbaug/git/OpenCL-CTS/test_conformance/api/negative_device.cpp:399)
```

Reply:
Hmm, the order in which error conditions should be checked and reported is not specified, maybe that's a discussion we should have. In that particular case, and with the spec as it stands, I agree that your implementation should be allowed to return `CL_DEVICE_PARTITION_FAILED`. The test could only check for that particular `CL_INVALID_VALUE` when there exists at least one valid partitioning scheme for the device. That would take care of the ambiguity.